### PR TITLE
Fixing a bug where model weights may be different on different workers.

### PR DESCRIPTION
**Pull request by Lucaweihs**

### CODE REVIEW

File: allenact/algorithms/onpolicy_sync/engine.py [hunk 1]
```diff
@@ -155,24 +159,36 @@ def __init__(
...
+
+            set_seed(self.seed)
+            self.actor_critic = cast(
+                ActorCriticModel, self.config.create_model(**create_model_kwargs),
+            ).to(self.device)
+
+        if initial_model_state_dict is not None:
+            self.actor_critic.load_state_dict(state_dict=initial_model_state_dict)
+        else:
+            assert mode != "train" or self.num_workers == 1, (
```
Comment (jordis-ai2) [inline allenact/algorithms/onpolicy_sync/engine.py:179]: isn't it enough to pass a default seed to all models when more than 1 worker? runner could check whether a seed has been set, and, if not, generate a single random one for all workers. model params should be initialized with the same seed, wouldn't it?

File: allenact/algorithms/onpolicy_sync/runner.py [hunk 2]
```diff
@@ -50,16 +52,22 @@ def __init__(
...
         disable_tensorboard: bool = False,
         disable_config_saving: bool = False,
     ):
         self.config = config
         self.output_dir = output_dir
         self.loaded_config_src_files = loaded_config_src_files
-        self.seed = seed
+        self.seed = seed if seed is not None else random.randint(0, 2 ** 31 - 1)
         self.deterministic_cudnn = deterministic_cudnn
+        if multiprocessing_start_method == "default":
```
Comment (jordis-ai2) [inline allenact/algorithms/onpolicy_sync/runner.py:65]: nice

File: allenact/algorithms/onpolicy_sync/runner.py [hunk 3]
```diff
@@ -76,8 +84,7 @@ def __init__(
         if self.deterministic_cudnn:
             set_deterministic_cudnn()
 
-        if self.seed is not None:
-            set_seed(self.seed)
+        set_seed(self.seed)
```
Comment (jordis-ai2) [inline allenact/algorithms/onpolicy_sync/runner.py:87]: isn't it enough to pass a common seed to the train workers? (when we do that, all models are initialized equally, this was checked with the TODO code block in engine/update(...))

File: allenact/algorithms/onpolicy_sync/runner.py [hunk 4]
```diff
@@ -234,9 +245,15 @@ def start_train(
         devices = self.worker_devices("train")
         num_workers = len(devices)
 
-        seed = (
-            self.seed
-        )  # same for all workers. used during initialization of the model
+        trainer_seeds = OnPolicyRLEngine.worker_seeds(
+            nprocesses=num_workers, initial_seed=self.seed
+        )
+        set_seed(self.seed)
```
Comment (jordis-ai2) [inline allenact/algorithms/onpolicy_sync/runner.py:251]: as I said, this should be handled by engine unless things have changed :?

File: allenact/algorithms/onpolicy_sync/runner.py [hunk 5]
```diff
@@ -256,13 +273,14 @@ def start_train(
                     if self.running_validation
                     else None,
                     checkpoints_dir=self.checkpoint_dir(),
-                    seed=seed,
+                    seed=trainer_seeds[trainer_it],
```
Comment (jordis-ai2) [inline allenact/algorithms/onpolicy_sync/runner.py:276]: I think this is going to break. the original (shared) seed is used to initialize the model...

**Comment by Lucaweihs** [comment]
@jordis-ai2 made the updates as we discussed, i.e. keeping the model initialization in the runner but fixing things so that the same seed is passed to all trainers as before.

File: allenact/algorithms/onpolicy_sync/runner.py [hunk 6]
```diff
@@ -50,16 +52,22 @@ def __init__(
...
+        multiprocessing_start_method: str = "default",
         extra_tag: str = "",
         disable_tensorboard: bool = False,
         disable_config_saving: bool = False,
     ):
         self.config = config
         self.output_dir = output_dir
         self.loaded_config_src_files = loaded_config_src_files
-        self.seed = seed
+        self.seed = seed if seed is not None else random.randint(0, 2 ** 31 - 1)
```
Comment (jordis-ai2) [inline allenact/algorithms/onpolicy_sync/runner.py:63]: this is great. then we ensure we always restart with the same (randomly chosen) seed. it was the last thing I was a bit annoyed with (that I could catch). nice! 👍

**Review by jordis-ai2 — APPROVED** [review]
LGTM
